openscm/scmdataframebase.py

```diff
@@ -1,34 +1,504 @@
+import copy
+import os
 from datetime import datetime
+from logging import getLogger
+
+import dateutil
+import numpy as np
+import pandas as pd
 from dateutil import parser
+from pyam.core import _raise_filter_error, IamDataFrame
+from pyam.utils import (
+    isstr,
+    IAMC_IDX,
+    years_match,
+    month_match,
+    day_match,
+    datetime_match,
+    hour_match,
+    pattern_match,
+    to_int,
+)
+
+logger = getLogger(__name__)
+
+
+def read_pandas(fname, *args, **kwargs):
+    """Read a file and return a pd.DataFrame"""
+    if not os.path.exists(fname):
+        raise ValueError("no data file `{}` found!".format(fname))
+    if fname.endswith("csv"):
+        df = pd.read_csv(fname, *args, **kwargs)
+    else:
+        xl = pd.ExcelFile(fname)
+        if len(xl.sheet_names) > 1 and "sheet_name" not in kwargs:
+            kwargs["sheet_name"] = "data"
+        df = pd.read_excel(fname, *args, **kwargs)
+    return df
+
+
+def read_files(fnames, *args, **kwargs):
+    """Read data from a snapshot file saved in the standard IAMC format
+    or a table with year/value columns
+    """
+    if not isstr(fnames):
+        raise ValueError(
+            "reading multiple files not supported, "
+            "please use `pyam.IamDataFrame.append()`"
+        )
+    logger.info("Reading `{}`".format(fnames))
+    return format_data(read_pandas(fnames, *args, **kwargs))
+
+
+def format_data(df, **kwargs):
```

Comment:
do we need to copy this entire function out of pyam or is there a way to split `format_data` in pyam so we can not do the reshaping bit here?

Reply:
Fair, it's also impossible to sensibly split as `format_data` is such an enormous function in pyam. I'll make a PR into pyam to split it. If we want we could reduce duplication in future.